privacyidea/lib/monitoringmodules/sqlstats.py

```diff
@@ -0,0 +1,192 @@
+#
+__doc__ = """This module writes statistics data to the SQL database table "monitoringstats".
+"""
+import logging
+from privacyidea.lib.monitoringmodules.base import Monitoring as MonitoringBase
+from privacyidea.lib.pooling import get_engine
+from sqlalchemy.orm import sessionmaker, scoped_session
+from privacyidea.lib.utils import censor_connect_string, convert_timestamp_to_utc
+from privacyidea.lib.lifecycle import register_finalizer
+log = logging.getLogger(__name__)
+from sqlalchemy import MetaData, cast, String
+from sqlalchemy import asc, desc, and_, or_
+from privacyidea.models import MonitoringStats
+from sqlalchemy import create_engine
+from sqlalchemy.orm import sessionmaker, scoped_session
+import traceback
+from privacyidea.lib.tokenclass import AUTH_DATE_FORMAT
+from dateutil.tz import tzlocal, tzutc
+
+
+metadata = MetaData()
+
+
+class Monitoring(MonitoringBase):
+
+    def __init__(self, config=None):
+        self.name = "sqlstats"
+        self.config = config or {}
+
+        # We can use "sqlstats" as the key because the SQLAudit connection
+        # string is fixed for a running privacyIDEA instance.
+        # In other words, we will not run into any problems with changing connect strings.
+        self.engine = get_engine(self.name, self._create_engine)
+        # create a configured "Session" class. ``scoped_session`` is not
+        # necessary because we do not share session objects among threads.
+        # We use it anyway as a safety measure.
+        Session = scoped_session(sessionmaker(bind=self.engine))
+        self.session = Session()
+        # Ensure that the connection gets returned to the pool when the request has
+        # been handled. This may close an already-closed session, but this is not a problem.
+        register_finalizer(self.session.close)
+        self.session._model_changes = {}
+
+    def _create_engine(self):
+        """
+        :return: a new SQLAlchemy engine connecting to the database specified in PI_MONITORING_SQL_URI.
+        """
+        # an Engine, which the Session will use for connection
+        # resources
+        connect_string = self.config.get("PI_MONITORING_SQL_URI", self.config.get(
+            "SQLALCHEMY_DATABASE_URI"))
+        log.debug("using the connect string {0!s}".format(censor_connect_string(connect_string)))
+        try:
+            pool_size = self.config.get("PI_MONITORING_POOL_SIZE", 20)
+            engine = create_engine(
+                connect_string,
+                pool_size=pool_size,
+                pool_recycle=self.config.get("PI_MONITORING_POOL_RECYCLE", 600))
+            log.debug("Using SQL pool size of {}".format(pool_size))
+        except TypeError:
+            # SQLite does not support pool_size
+            engine = create_engine(connect_string)
+            log.debug("Using no SQL pool_size.")
+        return engine
+
+    def add_value(self, stats_key, stats_value, utc_timestamp, reset_values=False):
```

Comment:
I think there is a slight discrepancy between ``add_value``, ``delete`` and ``get_values``:
* Both ``delete`` and ``get_values`` expect timezone-aware datetime objects and call ``convert_timestamp_to_utc`` on them to convert them to timezone-naive UTC datetime objects. I think this makes sense.
* But ``add_value`` doesn't call ``convert_timestamp_to_utc`` on its argument. This currently works because ``write_stats`` is the only caller of ``add_value`` and  calls ``convert_timestamp_to_utc`` on its datetime argument *before* passing it to ``add_value``. But I think it would be more consistent if ``add_value`` also expects a timezone-aware datetime object.

So, in other words: I think we should move this line
https://github.com/privacyidea/privacyidea/blob/56c18ff57362bb1401ce54793621062385ce3220/privacyidea/lib/monitoringstats.py#L72
to ``add_value``.

Reply:
done in fa76b0b

Reply:
:+1: